gnosis/gp-v2-services
# *New HTTP Solver Model incorporating Balancer Pools

Closes #769.

This PR is the accumulation of several smaller changes that result in a complete refactoring of the HTTP solver request and response format to accommodate Balancer weighted pool amm models.

This is the accumulation of

- use H160 token addresses instead of String (#763)
- Prepare Balancer Pool Data for HTTP Solver (#769)
- Weighted Pool Model with Liquidity (#781)
- Weighted Pool Liquidity Model (#775)
- Move H160Wrapper to shared (#787)
- update response model (#833) 
- UpdatedAmmModel to new format (#808) 


To align with the request type specification in https://github.com/gnosis/gp-v2-solver/issues/147#issuecomment-864878822 and in preparation for encoding Balancer Pools we change the UniswapModel (passed on to the HTTP solver) to be more flexible for more general pool types.

Essentially, we have changed from having keys `token0, token1, balance0, balance1` to `reserves` as part of the pool "parameters" like so.

```rust
#[derive(Debug, Serialize)]
pub struct AmmModel {
    #[serde(flatten)]
    pub parameters: AmmParameters,
    pub cost: CostModel,
    pub mandatory: bool,
}

#[derive(Debug, Serialize)]
#[serde(tag = "kind")]
pub enum AmmParameters {
    ConstantProduct(ConstantProductPoolParameters),
    WeightedProduct(WeightedProductPoolParameters),
}
```

where constant product parameters are a `HashMap<Token, Balance>` and weighted product pool parameters are similar, but whose hashmap values are another struct holding the balance and weight of that token.

The response data is also generalized to incorporate the fact that a single amm (pool) may be used multiple times in a settlement meaning that we had to use a vector of execution plans.

### Test Plan
Introduced several new test for json (de)serialization of the top level struct `BatchAuctionModel` and response model `SettledBatchAuctionModel`.

A local Integration test was performed as follows

1. Run the orderbook - make sure to have the orderbook DB running (cf services readme)
```sh
cargo run --bin orderbook -- --node-url <YOUR_NODE_URL>
```
2. FrontEnd: Change API URL to local host and run
```sh
REACT_APP_API_URL_STAGING_MAINNET=http://0.0.0.0:8080/api yarn start
```

3. Run Solver: Note that Uniswap must be included in `baseline-sources` to get a price estimate for balancer pools.
```sh
cargo run -p solver -- --baseline-sources BalancerV2,Uniswap --node-url <YOUR_NODE_URL> --solvers Mip --mip-solver-url <HTTP_SOLVER_URL> --base-tokens '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'
```
